mlc_llm/transform/smoothquant.py

```diff
@@ -0,0 +1,245 @@
+                self.pattern = is_op("relax.matmul")(self.lhs_sm, self.permute)
+            
+            def transform(self) -> tvm.IRModule:
+                for gv, func in self.mod.functions.items():
+                    if not isinstance(func, relax.Function):
+                        continue
+                    self.var2val = tvm.relax.analysis.get_var2val(func)
+                    self.profile_points = []
+                    self.params_to_remove = []
+                    updated_func = self.visit_expr(func)
+                    updated_func = remove_all_unused(updated_func)
+                    self.builder_.update_func(gv, updated_func)
+                return self.builder_.get()
+            
+            def visit_function_(self, f):
+                body = super().visit_expr(f.body)
+                new_params = [param for param in f.params if param not in self.params_to_remove]
+                return relax.Function(new_params, body, None, f.is_pure, f.attrs, f.span)
+            
+            def visit_seq_expr_(self, op: relax.SeqExpr) -> relax.Expr:
+                op = super().visit_seq_expr_(op)
+                if len(self.profile_points) != 0:
+                    new_body = relax.Tuple([op.body, relax.Tuple(self.profile_points)])
+                    return relax.SeqExpr(op.blocks, new_body, op.span)
+                return op
+
+            def visit_dataflow_block_(self, block: relax.DataflowBlock) -> relax.DataflowBlock:
+                self.builder_._begin_dataflow_block()
+                for binding in block.bindings:
+                    self.visit_binding(binding)
+                if len(self.profile_points) != 0:
+                    self.builder_.emit_output(self.profile_points)
+                return self.builder_._end_block()
+
+            def visit_call_(self, call: relax.Call) -> relax.Expr:
+                call = super().visit_call_(call)
+                matchings = self.pattern.extract_matched_expr(call, self.var2val)
+                if matchings:
+                    m_smq1 = matchings[self.lhs_sm]
+                    m_smq2 = matchings[self.rhs_sm]
+                    a_info = matchings[self.a_scale].struct_info
+                    w_info = matchings[self.w_scale].struct_info
+                    a_scale = tvm.runtime.ndarray.empty(a_info.shape, a_info.dtype, tvm.cpu())
+                    w_scale = tvm.runtime.ndarray.empty(w_info.shape, w_info.dtype, tvm.cpu())
+                    a_smq = self.builder_.emit(
+                        R.smooth(m_smq1.args[0], relax.Constant(a_scale), kind=1, mode="identity")
+                    )
+                    w_smq = self.builder_.emit(
+                        R.smooth(m_smq2.args[0], relax.Constant(w_scale), kind=2, mode="identity")
+                    )
+                    a_out = self.builder_.emit(R.absmax(a_smq, kind=1), "a_out")
+                    w_out = self.builder_.emit(R.absmax(w_smq, kind=2), "w_out")
```

Comment:
I'm seeing many `max` applied in the python side `smoothquant_utils.py` as well.

1. Can you document the purpose of each max where it is used?
2. If we want to support per-channel or per-token quantization, which max needs to be modified?

Reply:
A few words about `R.absmax`:

1. It was added for optimization purpose only. I need to dump tensor for statistics collection. But for large models this memory footprint is **huge**! That's why I decided to introduce this op and dump already preprocessed tensors.
2. Now this op operates like `R.abs() --> R.max() --> R.squeeze()`. I can use this sequence of ops instead of new one.
3. About per-channel or per-token quantization: yes, you are right. In this case it does not fit. We need to modify this op or add new one.

For current quantization scheme (per-tensor) we can avoid using `R.absmax` and use `R.abs() --> R.max() --> R.squeeze()` instead. For per-channel or per-token quantization we need extra ops for preprocessing. 
In my opinion dump the whole tensors is not very good idea and we need to preprocess it before dump (to optimize memory footprint)

Reply:
If we are dumping max from cpp, why do we need `np.max` on the python side?

Reply:
And can you tell me why having `R.absmax` helps memory for per-channel or per-token cases, compared to a chain of plain ops?

Reply:
About `np.max:`
I can not get rid of it, because it is used to merge calibration statistics from different elements in dataset.
For example:
We have 3 prompts in our dataset - prompt1, prompt2, prompt3. For each prompt we have its own statistics: stat1, stat2, stat3. To calculate final smoothing/quantization parameters I need to merge this stat: stat1, stat2, stat3 --> final_stat and work with this joined information. During this "merge" `np.max` is used.

Reply:
About `R.absmax`:
`R.absmax` should be identical to the chain of plain ops. This op does not reduce memory footprint compared to this chain (`R.abs, R.max, R.squeeze`).

But `R.absmax` itself helps to reduce memory in the following way, example:
Given `R.linear` op has the following input data (activations, 2D tensor): 
[[10, 1, 1],
 [1, 20, 1],
 [1, 1, 30]]
I do not save the whole tensor, after `R.absmax` I get 1D tensor with 3 elements: [10, 20, 30]. And I dump only this 1D vector. As a result, memory reduce is 3x times (9 vs 3 elements).

This toy example demonstrates how R.absmax can help reduce memory consumption.